src/python/pants/util/value_interpolation.py

```diff
@@ -0,0 +1,109 @@
+        msg = f"The placeholder {attribute!r} is unknown."
+        if value and isinstance(value, InterpolationValue):
+            msg += f' Try with one of: {", ".join(value.keys())}.'
+        return cls(msg)
+
+
+ErrorT = TypeVar("ErrorT", bound=InterpolationError)
+
+
+class InterpolationValue(FrozenDict[str, str]):
+    """Dict class suitable for use as a format string context object, as it allows to use attribute
+    access rather than item access."""
+
+    _attribute_error_type: ClassVar[type[InterpolationError]] = InterpolationError
+
+    def __getattr__(self, attribute: str) -> str:
+        if attribute not in self:
+            raise self._attribute_error_type.attribute_error(self, attribute)
+        return self[attribute]
+
+
+class InterpolationContext(FrozenDict[str, Union[str, InterpolationValue]]):
+    @classmethod
+    def from_dict(cls, data: Mapping[str, str | Mapping[str, str]]) -> InterpolationContext:
+        return InterpolationContext({key: cls.create_value(value) for key, value in data.items()})
+
+    @staticmethod
+    def create_value(value: str | Mapping[str, str]) -> str | InterpolationValue:
+        """Ensure that `value` satisfies the type `InterpolationValue`."""
+        if isinstance(value, (str, InterpolationValue)):
+            return value
+        return InterpolationValue(value)
+
+    def merge(self, other: Mapping[str, str | Mapping[str, str]]) -> InterpolationContext:
+        return InterpolationContext.from_dict({**self, **other})
+
+    def format(
+        self, text: str, *, source: TextSource, error_cls: type[ErrorT] | None = None
+    ) -> str:
+        stack = [text]
+        try:
+            while "{" in stack[-1] and "}" in stack[-1]:
+                if len(stack) >= 5:
+                    raise InterpolationError(
+                        "The formatted placeholders recurse too deep.\n"
+                        + " => ".join(map(repr, stack))
+                    )
+                stack.append(stack[-1].format(**self))
+                if stack[-1] == stack[-2]:
+                    break
+            return stack[-1]
+        except (KeyError, InterpolationError) as e:
+            default_error_cls = InterpolationError
+            msg = f"Invalid value for the {source}: {text!r}.\n\n"
+            if isinstance(e, InterpolationError):
+                default_error_cls = type(e)
+                msg += str(e)
+            else:
+                # KeyError
+                msg += f"The placeholder {e} is unknown."
+                if self:
+                    msg += f" Try with one of: {', '.join(sorted(self.keys()))}."
+                else:
+                    msg += (
+                        " There are currently no known placeholders to use. These placeholders "
+                        "can come from `[docker].build_args` or parsed from instructions in your "
+                        "`Dockerfile`."
```

Comment:
This part of the message needs to come from outside this class now, I think.

Reply:
what if we point the user to check the documentation of the `source` (whether it is an option scope or a target field)?

I'll push an update to better show what I mean.